crates/energy/src/graph.rs

```diff
@@ -0,0 +1,242 @@
+use std::collections::HashSet;
+
+use bevy::prelude::*;
+use bevy::utils::petgraph::prelude::*;
+use bevy::utils::petgraph::visit::IntoNodeReferences;
+use bevy::utils::Instant;
+use bevy_prototype_debug_lines::{DebugLines, DebugLinesPlugin};
+use de_core::gamestate::GameState;
+use de_core::objects::Active;
+use de_core::projection::ToFlat;
+use de_index::SpatialQuery;
+use de_spawner::{DespawnedComponentsEvent, DespawnEventsPlugin, SpawnerSet};
+use parry3d::bounding_volume::Aabb;
+use parry3d::math::Point;
+use smallvec::{smallvec, SmallVec};
+
+// The max distance (in meters) between two entities for them to be consider neighbors in the graph
+const MAX_DISTANCE: f32 = 10.0;
+
+pub(crate) struct GraphPlugin;
+
+impl Plugin for GraphPlugin {
+    fn build(&self, app: &mut App) {
+        app.add_plugins((DebugLinesPlugin::default(), DespawnEventsPlugin::<&NearbyUnits, NearbyUnits>::default()))
+            .add_systems(OnEnter(GameState::Playing), setup)
+            .add_systems(OnExit(GameState::Playing), clean_up)
+            .add_systems(Update, spawn_graph_components.after(SpawnerSet::Spawn))
+            .add_systems(
+                PreUpdate,
+                (
+                    remove_old_nodes.before(GraphSystemSet::UpdateNearby),
+                    update_nearby_recv.in_set(GraphSystemSet::UpdateNearby),
+                    update_graph.in_set(GraphSystemSet::UpdateGraph).after(GraphSystemSet::UpdateNearby),
+                    debug_lines.after(GraphSystemSet::UpdateGraph),
+                )
+                    .run_if(in_state(GameState::Playing)),
+            );
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, SystemSet)]
+enum GraphSystemSet {
+    UpdateNearby,
+    UpdateGraph,
+}
+
+/// The power grid resource is used to store the power grid graph.
+#[derive(Resource, Debug, Clone)]
+pub(crate) struct PowerGrid {
+    /// The power grid graph.
+    graph: GraphMap<Entity, f64, Undirected>,
+}
+
+impl Default for PowerGrid {
+    fn default() -> Self {
+        Self {
+            graph: GraphMap::new(),
+        }
+    }
+}
+
+/// The energy receiver component is used to mark an entity as an energy receiver.
+#[derive(Component, Debug, Clone, Copy)]
+pub struct EnergyReceiver;
+
+/// The energy producer component is used to mark an entity as an energy producer.
+#[derive(Component, Debug, Clone, Copy)]
+pub struct EnergyProducer;
+
+/// The nearby component is used to store the nearby entities of an entity.
+#[derive(Debug, Clone)]
+pub enum Nearby {
+    Receiver(SmallVec<[Entity; 40]>),
+    Producer(SmallVec<[Entity; 40]>),
+}
+
+impl Nearby {
+    fn into_inner(self) -> SmallVec<[Entity; 40]> {
+        match self {
+            Nearby::Receiver(inner) => inner,
+            Nearby::Producer(inner) => inner,
+        }
+    }
+}
+
+#[derive(Component, Default, Debug, Clone)]
+pub struct NearbyUnits(SmallVec<[Nearby; 2]>, Option<Vec2>);
```

Comment:
* Named fields would make the code easier to read.
* Why isn't this `{ procucers: SmallVec<[Entity; 40]>, receivers: SmallVec<[Entity; 40]>, ... }`?

Reply:
You mentioned at some point that nearby would be useful for many different reasons (attacking, ai, etc). And by using an enum it means that this can be extended with many different categories. The final resting place for the nearby logic should be somewhere like `de_index`.

Reply:
You are right, my guess is that we will utilize it in other systems as well. However, until we do, I would keep the logic as simple as possible.

Reply:
How about I split this logic into a new PR and put it into de_index?

Reply:
I think that we should keep the logic highly specialized for the energy until we encounter first case where we would utilize a generalized logic. At that point, we can move it to `de_index` and make it somewhat more general. YAGNI